### PR TITLE
fix: modern jest timers

### LINKED ISSUE #391: Version 2.x not compatible with jest.useFakeTimers('modern');
## Describe the bug

Using `jest.useFakeTimers('modern');` in any spec file, causes the tests to fail due to timeout error:

> Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error: Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.

Downgrading to 1.14 fixes this, therefore I believe this is related to 2.x. 

I am using  `jest.useFakeTimers('modern');` in order to call `jest.setSystemTime` in my test just FYI.

## Expected behavior

Tests should pass

## Steps to Reproduce

Simple component:
```
import React from 'react';
import { Text } from 'react-native';

const TestComponent = () => (
  <Text>hello world!</Text>
);

export default TestComponent;

```

spec file:
```
import React from 'react';
import { render } from 'react-native-testing-library';

import TestComponent from 'src/components/TestComponent';

jest.useFakeTimers('modern');

describe('TestComponent', () => {
  it('should render hello world!', () => {
    const utils = render(<TestComponent />);

    expect(utils.getByText('hello world!')).not.toBeNull();
  });
});
```

## Versions

```
    react: 16.13.1 => 16.13.1 
    react-native: 0.62.2 => 0.62.2 
    react-native-testing-library: ^2.1.0 => 2.1.0 
    react-test-renderer: ^16.13.0 => 16.13.1 
```

### PR DESCRIPTION
**Pull request by thymikee**
### Summary

Fixes #391 

### Test plan

Added regression tests.

### CODE REVIEW

File: src/__tests__/auto-cleanup.js [hunk 1]
```diff
@@ -25,9 +25,22 @@ class Test extends React.Component<*> {
...
+  expect(isMounted).toEqual(true);
   expect(fn).not.toHaveBeenCalled();
 });
 
 test('component is automatically umounted after first test ends', () => {
   expect(isMounted).toEqual(false);
 });
+
+test('does not time out with fake timers', () => {
+  jest.useFakeTimers();
```
Comment (mdjastrzebski) [inline src/__tests__/auto-cleanup.js:37]: As I understand there are 3 possible values for `useFakeTimers`: `'modern'`, `'legacy'` & default. With the default changing between Jest versions. Maybe we could have 3 tests here, default, legacy & modern, WDYT?

File: src/__tests__/waitFor.test.js [hunk 2]
```diff
@@ -87,9 +91,24 @@ test('works with fake timers', async () => {
...
     // suppress
   }
-  jest.runTimersToTime(400);
+  jest.advanceTimersByTime(400);
 
   expect(mockFn).toHaveBeenCalledTimes(3);
+});
 
-  jest.useRealTimers();
+test('works with modern fake timers', async () => {
```
Comment (mdjastrzebski) [inline src/__tests__/waitFor.test.js:99]: Similar as in autocleanup: test functions for default, modern and legacy timers.

File: src/flushMicroTasks.js [hunk 3]
```diff
@@ -2,14 +2,33 @@
...
-  return new Promise((resolve) => setImmediate(resolve));
+let enqueueTask;
+try {
+  // assuming we're in node, let's try to get node's
+  // version of setImmediate, bypassing fake timers if any.
+  // $FlowFixMe - timers is internal Node module
+  enqueueTask = require('timers').setImmediate;
+} catch (_err) {
+  // we're in a browser, do nothing
+  enqueueTask = (resolve) => {
```
Comment (mdjastrzebski) [inline src/flushMicroTasks.js:23]: How can we run in a browser?

File: src/flushMicroTasks.js [hunk 4]
```diff
@@ -2,14 +2,33 @@
 
 import { printDeprecationWarning } from './helpers/errors';
 
+type SimpleThenable = { then: (() => mixed) => mixed };
```
Comment (mdjastrzebski) [inline src/flushMicroTasks.js:5]: Consider renaming to `Thenable`

Comment (mdjastrzebski) [reply src/flushMicroTasks.js:5]: Consider making this us a generic type param

File: src/__tests__/auto-cleanup-skip.test.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ import { View } from 'react-native';
 let render;
 beforeAll(() => {
   process.env.RNTL_SKIP_AUTO_CLEANUP = 'true';
-  const rtl = require('../');
+  const rtl = require('..');
   render = rtl.render;
```
Comment (mdjastrzebski) [inline src/__tests__/auto-cleanup-skip.test.js:8]:
```suggestion
  const rntl = require('..');
  render = rntl.render;
```

**Comment by pke** [comment]
I wanted to report, that modern fake timers still (again?) don't work in `@testing-library/react-native@7.1.0`
When enabled all tests run into timeouts.

```tsx
test("renders greeting depending on time", () => {
  jest.useFakeTimers("modern")
  jest.setSystemTime(new Date("2020-11-19T09:00:00.000+01:00"))
  const { getByText } = render(<Greeting/>)
  expect(getByText("Guten Tag")).toBeTruthy()
})
```

Uncommenting the first 2 lines in the test make it work (but now always pass due to its dependency on current time).

**Comment by thymikee** [comment]
@pke we're aware of the issue and tracking it here: https://github.com/callstack/react-native-testing-library/issues/506 Feel free to help reviewing and testing the ongoing work: https://github.com/callstack/react-native-testing-library/pull/568 :)

**Pull request merged**